include/deal.II/differentiation/ad/ad_drivers.h

```diff
@@ -0,0 +1,1270 @@
+     */
+    template<typename ADNumberType, typename ScalarType, typename T = void>
+    struct TapedDrivers
+    {
+      // This dummy class definition safely supports compilation
+      // against tapeless numbers or taped number types that have
+      // not yet been implemented.
+
+      /**
+       * @name Taping
+       */
+      //@{
+
+      /**
+       * Enable the recording mode for a given tape.
+       *
+       * @param[in] tape_index The index of the tape to be written
+       * @param[in] keep_independent_values Determines whether the numerical
+       *            values of all independent variables are recorded in the
+       *            tape buffer.
+       */
+      static void
+      enable_taping(const unsigned int &tape_index,
+                    const bool         &keep_independent_values);
+
+      /**
+       * Disable the recording mode for a given tape.
+       *
+       * @param[in] active_tape_index The index of the (currently active) tape
+       *            to be finalized and potentially written to file.
+       * @param[in] write_tapes_to_file A flag that specified whether the tape
+       *            should be written to file or kept in memory.
+       */
+      static void
+      disable_taping(const unsigned int &active_tape_index,
+                     const bool         &write_tapes_to_file);
+
+      /**
+       * Prints the statistics regarding the usage of the tapes.
+       *
+       * @param[in] stream The output stream to which the values are to be written.
+       * @param[in] tape_index The index of the tape to get the statistics of.
+       */
+      template<typename Stream>
+      static void
+      print_tape_stats(Stream             &stream,
+                       const unsigned int &tape_index);
+
+      //@}
+
+      /**
+       * @name Drivers for scalar functions (one dependent variable)
+       */
+      //@{
+
+      /**
+       * Computes the value of the scalar field.
+       *
+       * @param[in] active_tape_index The index of the tape on which the dependent
+       *            function is recorded.
+       * @param[in] independent_variables The scalar values of the independent
+       *            variables whose sensitivities were tracked.
+       *
+       * @return The scalar values of the function.
+       */
+      static ScalarType
+      value (const unsigned int            &active_tape_index,
+             const std::vector<ScalarType> &independent_variables);
+
+      /**
+       * Computes the gradient of the scalar field with respect to all independent
+       * variables.
+       *
+       * @param[in] active_tape_index The index of the tape on which the dependent
+       *            function is recorded.
+       * @param[in] independent_variables The scalar values of the independent
+       *            variables whose sensitivities were tracked.
+       * @param[out] gradient The scalar values of the dependent function's gradients.
```

Comment:
Does the gradient variable have to have the correct size already, or is it resized?

Same below for the Hessian matrix.